### PR TITLE
Add PowerSaves for 3DS device support — DS cartridge save backup

### PR DESCRIPTION
**Pull request by pathawks**
PowerSaves for 3DS is a USB HID slot-1 adapter (VID `0x1C1A` PID `0x03D5`) made by Datel. Despite the 3DS branding, the device's generic NTR and SPI passthroughs work cleanly on DS cartridges, so this driver uses it purely as a save-backup tool for DS carts.

## What ships

- **Driver** (`src/lib/drivers/powersave-3ds/`) — speaks the 64-byte HID protocol, switches between ROM and SPI modes, identifies the cart via NTR chip ID, validates the on-cart header (CRC + boot logo), probes the save chip (FLASH via JEDEC, EEPROM via wrap-detection), and dumps the save with cart-swap detection at start and end. Protocol ported from [kitlith/powerslaves](https://github.com/kitlith/powerslaves) (MIT).
- **Shared NDS system** (`src/lib/systems/nds/`) — header parsing with CRC-16/MODBUS validation, maker/licensee code lookup (derived from [devkitPro/ndstool](https://github.com/devkitPro/ndstool), GPL-3.0), and a save-only system handler with generic dump-quality sanity checks. The header/family classifier and shared scanner UI are set up so future NDS-capable drivers can slot in without reshaping the layer.
- **Scanner UI** (`src/components/wizard/nds-scanner.tsx`, `src/hooks/use-nds-scanner.ts`, `src/components/shared/{cart-heading,nds-cart-info,save-dump-result}.tsx`) — polls for a cart, displays detected game info enriched with No-Intro lookup by serial, runs the dump, and surfaces hashes + `.sav` download. Post-dump panel mirrors `CompleteStep`'s status badge.
- **Wiring** — `App.tsx` routes `nds_save` devices to the new scanner; `connection-registry.ts` adds the device entry; `devices.ts` registers it; `linux/99-nabu.rules` grants hidraw access.

## Adjacent changes

- `nointro.ts` — add `nds_save` No-Intro system mapping; prefer retail entries over `(Beta)`/`(Proto)`/`(Demo)`/`(Sample)`/`(Unl)` tags when multiple entries share a serial, so `lookupBySerial` returns the cart users actually have.
- `hashing.ts` — `formatBytes` now scales to MB for ROM-size displays.
- `types.ts` — declare `nds_save` as a known `SystemId`; `CartridgeInfo` is now generic in its `meta` field; `DeviceDriver` gets an optional `dispose()` hook that `use-connection`'s `handleDisconnect` now calls.
- `alert.tsx` — add a `warning` variant for amber save-quality alerts.

## Tests

Vitest is added as a dev dependency (`npm run test`). Initial coverage:

- `crc16Modbus`, `parseNDSHeader`, and `classifyNDSCart` (success, corruption, and boundary cases)
- Prerelease-vs-retail serial preference in `NoIntroVerificationDB`
- `eepromWrapsAt` SPI boundary-detection helper

## Attribution

Sources cited in `THIRD-PARTY-LICENSES`: `kitlith/powerslaves` and `devkitPro/ndstool`. The `README` supported-hardware table is updated and includes a Datel trademark disclaimer.

### CODE REVIEW

File: src/lib/drivers/powersave-3ds/powersave-3ds-driver.ts [hunk 1]
```diff
@@ -0,0 +1,710 @@
...
+          "warn",
+        );
+        this.currentMode = null;
+        this.romInited = false;
+        await this.ensureRomInit();
+      }
+      this.header = parsed;
+      this.headerChipId = chipIdHex;
+      this.saveSize = 0;
+    }
```
[the remark this answers is not in the document]
Comment (pathawks) [reply src/lib/drivers/powersave-3ds/powersave-3ds-driver.ts:245]: Already addressed in 1e6fe8f from the first review round — `this.saveTypeName = undefined;` is set alongside `this.saveSize = 0;` in both the no-cart branch (current line 208) and the cart-swap branch (current line 244). `git blame` confirms the fix landed before this review was generated.

File: src/lib/drivers/powersave-3ds/powersave-3ds-commands.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+} as const;
+
+/**
+ * Decode FLASH capacity byte (third byte of JEDEC response) to size in bytes.
+ * Common NDS save-FLASH chips:
+ *   0x13 = 512 KB   0x14 = 1 MB   0x15 = 2 MB   0x16 = 4 MB
+ */
+export function flashSizeFromJedec(capacityByte: number): number | null {
+  if (capacityByte < 0x10 || capacityByte > 0x1f) return null;
+  return 1 << capacityByte;
```
[the remark this answers is not in the document]
Comment (pathawks) [reply src/lib/drivers/powersave-3ds/powersave-3ds-commands.ts:70]: `flashSizeFromJedec` caps the input at `0x18` (`if (capacityByte < 0x10 || capacityByte > 0x18) return null;`) so the shift never exceeds `1 << 24` = 16 MB, well inside the int32 range. The unit test at `commands.test.ts:17` exercises the boundary (`0x18` accepted, `0x19` and `0x1f` return null). No code change needed.

File: src/lib/drivers/powersave-3ds/powersave-3ds-driver.ts [hunk 3]
```diff
@@ -0,0 +1,794 @@
...
+      } else {
+        // Header returned non-0xFF data that failed CRC validation. The
+        // most likely cause is a 3DS cart returning encrypted bytes on
+        // the DS header-read path. The save chip sits on a separate SPI
+        // bus and doesn't participate in cart-bus encryption, so the
+        // save dump can still work. Could also be a DS cart with dirty
+        // contacts — same recovery: try the dump.
+        this.log(
+          "Header failed CRC validation — attempting save dump anyway.",
+          "warn",
```
[the remark this answers is not in the document]
Comment (pathawks) [reply src/lib/drivers/powersave-3ds/powersave-3ds-driver.ts]: Addressed in 561025e — the all-0xFF (3DS-signature) branch now throws `UnsupportedCartError` instead of proceeding to the SPI probe. The cart card stays visible via the wizard's `phase === 'error'` branch with a 'use a 3DS-cart-aware tool' message.

File: src/lib/drivers/powersave-3ds/powersave-3ds-driver.ts [hunk 4]
```diff
@@ -0,0 +1,794 @@
...
+
+    // 2-byte addressing path: 4 / 64 / 512 Kbit EEPROM.
+    const base = await this.spiReadAddr(0, 16, 2);
+    if (base.every((b) => b === base[0])) {
+      this.log(
+        "Save-chip size could not be auto-detected (save may be empty); " +
+          "defaulting to 64 KB (512 Kbit EEPROM).",
+        "warn",
+      );
+      return { size: 65536, typeName: "EEPROM (assumed)", addrWidth: 2 };
```
[the remark this answers is not in the document]
Comment (pathawks) [reply src/lib/drivers/powersave-3ds/powersave-3ds-driver.ts:610]: Deferred to issue #12 — needs a UI control (Save Size picker that surfaces when auto-detect falls back) which is bigger than a Copilot fix-it. Scope and code pointers captured there.

File: src/lib/systems/nds/nds-header.ts [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+ *   - title[0] is alphanumeric (real titles don't start with punctuation)
+ *   - title has ≥ 3 alphanumeric characters
+ *   - gameCode (0x0C..0x0F) is all alphanumeric
+ *   - gameCode[3] is a known NDS region letter
+ *   - capacity byte (0x14) is ≤ 0x0F (real carts are 3..12)
+ *
+ * Together these reject the nondeterministic preamble bytes some
+ * firmwares return before the real header.
+ */
+export function findHeaderStart(raw: Uint8Array): number {
```
[the remark this answers is not in the document]
Comment (pathawks) [reply src/lib/systems/nds/nds-header.ts:123]: Addressed in 50bd024 — `nds-header.test.ts` now has a `findHeaderStart` describe block with five cases: offset 0, single-byte preamble (the PowerSaves observed case), multi-byte preamble, garbage-buffer rejection, and unknown-region-letter rejection.

File: src/lib/drivers/powersave-3ds/powersave-3ds-driver.ts [hunk 6]
```diff
@@ -0,0 +1,794 @@
...
+   * asynchronously via the HID input listener. If we clear the inbox
+   * at the START of the next command, there's a race: stragglers in
+   * flight may land AFTER our clear, contaminating the new response's
+   * leading bytes as a fake "preamble" that looks like cart data but
+   * isn't.
+   *
+   * Fix: drain at both ends. Before sending, wait briefly for any
+   * in-flight stragglers to land, THEN clear. After receiving the
+   * requested bytes, wait briefly again and discard anything extra.
+   * That way the inbox boundary is clean when the next caller starts.
```
[the remark this answers is not in the document]
Comment (pathawks) [reply src/lib/drivers/powersave-3ds/powersave-3ds-driver.ts]: Addressed in 50bd024 — rewrote the docstring to describe the actual start-of-call drain. Stragglers from a given call get cleared by the *next* caller's start-of-call drain, which has proven sufficient. The post-receive drain idea was tried and dropped earlier in development.

File: src/lib/systems/nds/nds-save-system-handler.ts [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+      if (data[0] === 0) {
+        warnings.push(
+          "Dump is all zero bytes. The cartridge may have been blank " +
+            "(brand-new save), the contacts may be dirty, or the read " +
+            "may have failed. Re-seat the cartridge and dump again to compare.",
+        );
+      } else if (data[0] === 0xff) {
+        warnings.push(
+          "Dump is all 0xFF — the save chip didn't respond and the bus " +
+            "stayed idle. Re-seat the cartridge and dump again.",
```
[the remark this answers is not in the document]
Comment (pathawks) [reply src/lib/systems/nds/nds-save-system-handler.ts]: Addressed in 50bd024 — the all-0xFF warning now mirrors the all-0x00 one's wording, covering the blank/erased case (FLASH chips ship erased to 0xFF) alongside the bus-idle interpretation.

File: src/lib/core/nointro.ts [hunk 8]
```diff
@@ -165,6 +184,7 @@ export const NOINTRO_SYSTEM_NAMES: Readonly<Record<string, readonly string[]>> =
     gba: ["Nintendo - Game Boy Advance", "Game Boy Advance"],
     nes: ["Nintendo - Nintendo Entertainment System", "NES"],
     snes: ["Nintendo - Super Nintendo Entertainment System", "SNES"],
+    nds_save: ["Nintendo - Nintendo DS", "DS"],
```
[the remark this answers is not in the document]
Comment (pathawks) [reply src/lib/core/nointro.ts:187]: Addressed in d6dea3c — extracted a `matchesSystemName(datName, candidate)` helper in nointro.ts that requires exact match OR the No-Intro variant pattern (`candidate + ' ('`). Tests in nointro.test.ts verify the matcher rejects DSi against the DS alias, 3DS against the bare DS fallback, and Game Boy Color / Advance against the Game Boy alias.

File: src/lib/systems/nds/nds-save-system-handler.ts [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+
+  validate(_values: ConfigValues): ValidationResult {
+    return { valid: true };
+  }
+
+  buildReadConfig(values: ConfigValues): ReadConfig {
+    return {
+      systemId: "nds_save",
+      params: {
+        saveSize: values.saveSizeBytes as number | undefined,
```
[the remark this answers is not in the document]
Comment (pathawks) [reply src/lib/systems/nds/nds-save-system-handler.ts]: Addressed in 168aecc — renamed `params.saveSize` to `params.saveSizeBytes` to match the GB / GBA convention. No consumers currently read the field (PowerSaves 3DS `readROM` explicitly voids its config arg), so this is a forward-looking consistency fix.

File: src/lib/drivers/powersave-3ds/powersave-3ds-driver.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (pathawks) [reply src/lib/drivers/powersave-3ds/powersave-3ds-driver.ts]: Addressed in 561025e — the two cases are now split: all-0xFF DS-format header (the 3DS signature) is rejected with `UnsupportedCartError`, while the CRC-fail branch (dirty contacts / counterfeit-encrypted) still proceeds with a warning. Testing showed a real ST 512 KB FLASH dump recovered cleanly from a cart whose header read returned all zeros, so eager bail on CRC fail would prevent recoverable cases.

File: src/lib/drivers/powersave-3ds/powersave-3ds-driver.ts [hunk 10]
```diff
@@ -0,0 +1,794 @@
...
+      throw new UnsupportedCartError(
+        "This cart's save chip is behind an IR module that the " +
+          "PowerSaves for 3DS firmware cannot bypass — save dumping " +
+          "with this device is not possible. Use a different DS-cart " +
+          "adapter (EMS NDS Adapter, Neoflash SMS4) to back up this " +
+          "cart's save.",
+      );
+    }
+
+    if (jedecResp.some((b) => b !== 0x00 && b !== 0xff)) {
```
[the remark this answers is not in the document]
Comment (pathawks) [reply src/lib/drivers/powersave-3ds/powersave-3ds-driver.ts]: Addressed in 50bd024 — tightened the FLASH-path guard from `some` over all 3 bytes to `jedecResp[0] !== 0x00 && jedecResp[0] !== 0xff`. Real JEDEC manufacturer codes start at 0x01; the 0x00 and 0xFF cases now correctly fall through to the EEPROM probe path (or the IR-module detection ahead of it).

File: src/components/wizard/nds-scanner.tsx [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+        </Alert>
+      )}
+
+      {/* Polling state */}
+      {(phase === "polling" || phase === "idle") && (
+        <Card>
+          <CardContent className="flex items-center gap-3 py-8">
+            <span className="inline-block h-4 w-4 animate-spin rounded-full border-2 border-primary border-t-transparent" />
+            <span className="text-sm text-muted-foreground">
+              Insert a DS cartridge...
```
[the remark this answers is not in the document]
Comment (pathawks) [reply src/components/wizard/nds-scanner.tsx:89]: Addressed in fef0f23 — narrowed the ConnectStep safety text from the blanket 'always unplug before connecting/disconnecting a cartridge' to 'unplug before swapping cartridges; polled-insertion devices will prompt when ready.' The scanner's first-insertion-after-connect flow is intentional for PowerSaves and the other polled drivers (PS3 MCA, Disney Infinity, amiibo PowerSaves).
